core/shared/src/main/scala/fs2/text.scala

```diff
@@ -35,74 +38,119 @@ object text {
      * 0 is returned.
      */
     def lastIncompleteBytes(bs: Array[Byte]): Int = {
-      val lastThree = bs.drop(0.max(bs.size - 3)).toArray.reverseIterator
-      lastThree
-        .map(continuationBytes)
-        .zipWithIndex
-        .find {
-          case (c, _) => c >= 0
-        }
-        .map {
-          case (c, i) => if (c == i) 0 else i + 1
+      /*
+       * This is logically the same as this
+       * code, but written in a low level way
+       * to avoid any allocations and just do array
+       * access
+       *
+       *
+       *
+        val lastThree = bs.drop(0.max(bs.size - 3)).toArray.reverseIterator
+        lastThree
+          .map(continuationBytes)
+          .zipWithIndex
+          .find {
+            case (c, _) => c >= 0
+          }
+          .map {
+            case (c, i) => if (c == i) 0 else i + 1
+          }
+          .getOrElse(0)
+
+       */
+
+      val minIdx = 0.max(bs.size - 3)
+      var idx = bs.size - 1
+      var counter = 0
+      var res = 0
+      while (minIdx <= idx) {
+        val c = continuationBytes(bs(idx))
+        if (c >= 0) {
+          if (c == counter)
+            res = 0
+          else
+            res = counter + 1
+          // exit the loop
+          idx = 0
```

Comment:
If `bs.size` is `1` then what would happen? It seems like `minIdx` would be set to `0` (`0.max(-2)`) and `idx` would also be set to `0` (`1-1`). In that case, `idx = 0` won't break you out of the loop. I think `idx = -1` is a safer way to do this (although I would just use `return` myself).

Oh I see, you're counting on the decrement operator that occurs later. That makes sense, although IMO using `-1` or `return` here might be a bit more reassuring to the reader. (That said, there's no bug so maybe it's not a big deal.)